### ISSUE TITLE
Fix CTD

### ISSUE THREAD

**Issue by bschilder** [neurogenomics/MultiEWCE#14]
@NathanSkene @KittyMurphy
I just noticed that the CTD originally included within MultiEWCE is incorrect. It was not processed to the level of gene properly.

```R
MultiEWCE::load_example_ctd()
```
<img width="197" alt="Screenshot 2023-03-26 at 10 49 38" src="https://user-images.githubusercontent.com/34280215/227767880-d815078f-98dc-40ef-83fb-74e67fe5e3ee.png">

Instead, it appears to be at the level of transcripts, which were then randomly assigned unique gene-level IDs. You can also tell this right away when looking at the number of dimensions in the CTD. 
<img width="351" alt="Screenshot 2023-03-26 at 10 50 29" src="https://user-images.githubusercontent.com/34280215/227767961-9f875ee9-fce4-4abc-9d6f-f60786570d84.png">

@bobGSmith can you please confirm which version of the CTD you used in your original EWCE analyses? If it was this version, I will need to redo all of the analyses. Otherwise, our gene counts will be underestimated by ~2/3 and thus the celltype signatures would be degraded.

**Comment by NathanSkene** [comment]
Yikes, we’ll spotted Brian

Sent from Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: Brian M. Schilder ***@***.***>
Sent: Sunday, March 26, 2023 10:53:17 AM
To: neurogenomics/MultiEWCE ***@***.***>
Cc: Skene, Nathan G ***@***.***>; Mention ***@***.***>
Subject: [neurogenomics/MultiEWCE] Fix CTD (Issue #14)


This email from ***@***.*** originates from outside Imperial. Do not click on links and attachments unless you recognise the sender. If you trust the sender, add them to your safe senders list<https://spam.ic.ac.uk/SpamConsole/Senders.aspx> to disable email stamping for this address.



@NathanSkene<https://github.com/NathanSkene> @KittyMurphy<https://github.com/KittyMurphy>
I just noticed that the CTD originally included within MultiEWCE is incorrect. It was not processed to the level of gene properly.

MultiEWCE::load_example_ctd()

[Screenshot 2023-03-26 at 10 49 38]<https://user-images.githubusercontent.com/34280215/227767880-d815078f-98dc-40ef-83fb-74e67fe5e3ee.png>

Instead, it appears to be at the level of transcripts, which were then randomly assigned unique gene-level IDs. You can also tell this right away when looking at the number of dimensions in the CTD.
[Screenshot 2023-03-26 at 10 50 29]<https://user-images.githubusercontent.com/34280215/227767961-9f875ee9-fce4-4abc-9d6f-f60786570d84.png>

@bobGSmith<https://github.com/bobGSmith> can you please confirm which version of the CTD you used in your original EWCE analyses? If it was this version, I will need to redo all of the analyses. Otherwise, our gene counts will be underestimated by ~1/3 and thus the celltype signatures would be degraded.

**Comment by bschilder** [comment]
@bobGSmith could really use your input on this. Just need to know which version of the CTD you actually used to run the EWCE analyses.

**Comment by bschilder** [comment]
After discussing this with @bobGSmith I've confirmed that the issue stems from here, the script that was used to generate the Decartes (and Tabula Muris) CTDs:
https://github.com/neurogenomics/rare_disease_celltyping/blob/dev_descartes/source/CTD_Gen.R

@bobGSmith was given a version of this script @NathanSkene, as it had previously been written by Zijing. Not sure why this was suggested instead of the built-in `EWCE::generate_ctd` function (which should be have been available at the time), but the script doesn't account for a lot of variability in scRNA-seq dataset formats. In the case of Descartes, one of the consequences of this is not naming the genes correctly. Descartes provides their data at transcript level, so an extra steps must be done to aggregate to the data to gene level. it seems this was not done and so duplicated row names were automatically differentiated by R by appending ".1,",".2",".3", etc to each duplicated gene name.

In conclusion, I will indeed need to reprocess both CTDs and rerun all analyses for this project.

**Comment by bschilder** [comment]
# Assessing the issue 

Okay, so been going back and trying to reconstruct what happened. If I remember correctly @bobGSmith used the script provided by @NathanSkene because the merged Descartes dataset was too large for EWCE to process all at once. This was when I was still working on the `DelayedArray` implementation of `EWCE` which allows it to scale to an arbitrarily large dataset. So the strategy @bobGSmith took was to chunk each tissue-specific file from Descartes, aggregate them to cell-type level separately and then merge them back together into one CTD.

Another complication arose due to the way the genes in Descartes are annotated. As I mentioned, they appear to be annotated at the level of transcripts. Using the Gene Annotations file provided on their website, we can inspect this:
https://descartes.brotmanbaty.org/bbi/human-gene-expression-during-development/

```R
genes <- readRDS(url("https://atlas.fredhutch.org/data/bbi/descartes/human_gtex/downloads/data_summarize_fetus_data/df_gene.RDS"))
dim(genes)
# [1] 63561     3
```
![Screenshot 2023-04-12 at 13 13 24](https://user-images.githubusercontent.com/34280215/231453592-39fdf9d6-28b4-4d49-a389-5042bb32e3f3.png)

The first thing I noticed was the ".#" suffix at the end of each Ensembl gene ID. This usually indicates that the label is referring to a specific transcript of that gene. However, this suffix usually is given at the end of an Ensembl transcript ID (ENST) not a gene id (ENSG), which makes this annotation style quite confusing.
 For example, see the Transcript Table here:
https://grch37.ensembl.org/Homo_sapiens/Gene/TranscriptComparison?g=ENSG00000223972;r=1:11869-14412

Googling led me to this page from Illumina where they use the ENSG.# format:
https://support.illumina.com/content/dam/illumina-support/help/Illumina_DRAGEN_Bio_IT_Platform_v3_7_1000000141465/Content/SW/Informatics/Dragen/TPipelineGAFile_fDG.htm

The second thing I noticed was that there seemed to be a LOT of pseudogenes in Descartes. This may be related to the odd naming conventions I described.
```R
 table(genes$gene_type)|> sort(decreasing = TRUE)
```

![Screenshot 2023-04-12 at 13 22 02](https://user-images.githubusercontent.com/34280215/231455588-7a4fbf8d-776e-404c-8d0c-c1b2d52ccc2b.png)


Now, part of me is inclined to remove all these pseudogenes. However, "pseudogenes" are a bit of an open question in biology. They can indicate genes which have no function, or genes that have functions that we simply do not understand presently. Therefore, throwing them all out would make me a bit nervous, especially when they account for such a large proportion of the dataset.

Next, we can see that the number of unique ENSG Ids is only slightly larger than the total number of unique gene symbols provided. This probably have a lot to do with the large number of pseudogenes (and well as non-coding transcripts) included in this dataset. But even so, this seems like a crazy number of unique gene symbols. This makes me concerned that perhaps even the gene symbol naming conventions used by the Descartes team may or may not match up with the ones commonly used in other datasets:
```R
lapply(genes,function(x){length(unique(x))})
$gene_id
[1] 63561

$gene_type
[1] 30

$gene_short_name
[1] 56622
```
So @bobGSmith , your 59608 gene symbols is somewhere in between the number of unique ENGS IDs and the number of unique gene symbols in Descartes. I still suspect there may be artificially-induced gene symbol ".#" suffixes in your CTD given that the scripts weren't really designed for this scenario, but the consequences may not be quite as drastic as the ~2/3 gene signature degradation i originally estimated. Still, we should systematically address this to make sure we're processing the data correctly.

# How do we address this?

To explore this further, I tried doing some mapping with `orthogene`. This will give us a sense of how many of these Descartes-provided gene symbols exist in other databases:
```R
gp <- orthogene::all_genes(species="human",method = "gprofiler")
hg <- orthogene::all_genes(species="human",method = "homologene")
bg <- orthogene::all_genes(species="human",method = "babelgene")
```

![Screenshot 2023-04-12 at 13 31 01](https://user-images.githubusercontent.com/34280215/231457704-cd27c9e4-70f1-4472-ac8e-189eb741c8e3.png)

We can see there's a big discrepancy in the total number of genes available in each of these databases. This is likely bc homologene and babelgene are focused exclusively on converting gene *symbols* across *species*. While gprofiler  converts gene symbols/ensembl IDS/transcripts/proteins/etc within and across species.

So it would seem for the use case of Descartes, gprofiler would be most appropriate. But we still are left with the question of which gene annotation to use:
1. their ENSG.# notation.
2. their ENSG.# notation, aggregated after removing ".#" bit.
2. their ENSG.# notation, aggregated after removing ".#" bit AND standardising gene IDs using gprofiler.
4. aggregating the expression matrices (sum) using their gene symbol notation.
5. aggregating the expression matrices (sum) using their gene symbol notation AND standardising gene symbols using gprofiler.

Let's try some of these and see what we get:
```R

```

**Comment by bschilder** [comment]
Now that we've done this, we can check to see how many of these genes from each approach are in the HPO annotations, which is ultimately what we care about:

```R
gene_lists <- list(genes$gene_short_name, gene_map2$name, gene_map3$name)
d=HPOExplorer::load_phenotype_to_genes()
lapply(gene_lists, function(g){sum(unique(d$Gene) %in% g)})
```
![Screenshot 2023-04-12 at 14 26 09](https://user-images.githubusercontent.com/34280215/231471964-ce2a190b-9632-4fd1-98eb-551e4018dea0.png)

From this, we can see that `gene_map2` ("map their ENSG.# notation after removing ".#" bit and aggregate") gives us the best coverage of HPO genes. So I've decided that this is the best method for us to move forward with, and have reprocessed the CTD accordingy.

**Comment by bschilder** [comment]
Just to check how much this will affect our analyses we can compute some metrics for both the old and new CTDs.

We can compute the proportion of each phenotype's gene list that are present in each CTD.

## Old CTD

```R
ctd <- MultiEWCE::load_example_ctd()
ctd_genes <-  rownames(ctd[[1]]$mean_exp)])
pheno_counts <- d[,list(n_genes=length(unique(Gene[Gene %in% ctd_genes ),
                        total_genes=length(unique(Gene))),
                  by="HPO_ID"][,prop_genes:=n_genes/total_genes]
hist(pheno_counts$prop_genes)
```
![Screenshot 2023-04-12 at 15 17 08](https://user-images.githubusercontent.com/34280215/231486202-2b87c336-bd64-4352-8469-c0bf923e281c.png)
![Screenshot 2023-04-12 at 15 17 34](https://user-images.githubusercontent.com/34280215/231486314-11a162b0-d01d-4104-bf97-4ee0a6d2c48f.png)

## New CTD

```R
ctd_genes <-  rownames(ctd2[[2]]$mean_exp)
pheno_counts <- d[,list(n_genes=length(unique(Gene[Gene %in% ctd_genes]) ),
                        total_genes=length(unique(Gene))),
                  by="HPO_ID"][,prop_genes:=n_genes/total_genes]
hist(pheno_counts$prop_genes)
```
![Screenshot 2023-04-12 at 15 48 41](https://user-images.githubusercontent.com/34280215/231495486-1182aded-1127-4f38-96a2-6ca22a8017ba.png)
![Screenshot 2023-04-12 at 15 48 29](https://user-images.githubusercontent.com/34280215/231495515-cb315be9-3476-409f-bc54-c2f9bc063603.png)

We can see the new CTD increases the gene coverage to >98%, despite having fewer total genes than the old CTD (38259 vs. 59608). This is because the genes that remain in the new CTD have more generalisable gene symbols.
